File: LongestIncreasingSubsequence.java
package niuke;

//LIS练习题
public class LongestIncreasingSubsequence {
	public int getLIS(int[] A, int n) {
		int dp[] = new int[n];// dp[i] 表示最后一个字符为A[i]时的最长上升子序列的长度

		dp[0] = 1;

		int res = 0;
		for (int i = 1; i < n; i++) {

			int j = i;
			int max = 0;
			for (; j >= 0; j--) {// 每次从i位置向前遍历找到比i小的位置 取这些位置中子序列长度最长的作为上一个序列
				if (A[j] < A[i] && dp[j] > max) {
					max = dp[j];
				}
			}
			dp[i] = max + 1;
			res = Math.max(dp[i], res);// 记录整个过程中最长的上升子序列
		}

		return res;
	}
}

File: 第十二章_动态规划/LIS练习题/LongestIncreasingSubsequence.cpp
#include <iostream>
using namespace std;

const int N=500;

//dp[i]=k代表以数组i索引位置结尾的最长上升子序列长度为k 
int dp[N+1];

class LongestIncreasingSubsequence {
public:
    int getLIS(vector<int> A, int n) {
        dp[0]=1;//数组第一个数的最长上升子序列就是本身，所以为1
		int temp;
		int max=0;
		for(int i=1;i<n;i++){
			temp=0;
			for(int j=i-1;j>=0;j--){//从第i个数往前寻找比第i个数小的作为上升子序列，然后在这个过程中找到已知序列最长的那个。这样再加上第i个数本身使得最长子序列最长
				if(A[i]>A[j]&&dp[j]>temp){
					temp=dp[j];
				}
			}
			dp[i]=temp+1;
			if(dp[i]>max){
				max=dp[i];
			}
		}

		return max;
    }
};